### ISSUE TITLE
Comprehensive marble tests for filter

### ISSUE THREAD

**Issue by blesh** [ReactiveX/RxJS#390]
Should duplicate the current RxJS 3/4 tests

Should adhere to our contribution guidelines and operator authoring docs.

**Comment by staltz** [comment]
I'll start working on this. There are lots of tests here https://github.com/Reactive-Extensions/RxJS/blob/master/tests/observable/where.js

**Comment by staltz** [comment]
Most RxJS legacy tests for filter are now done. Except these:

- https://github.com/Reactive-Extensions/RxJS/blob/master/tests/observable/where.js#L288
- https://github.com/Reactive-Extensions/RxJS/blob/master/tests/observable/where.js#L608
- https://github.com/Reactive-Extensions/RxJS/blob/master/tests/observable/where.js#L668

**Comment by luisgabriel** [comment]
@staltz Rewriting the test this way it passes:

```javascript
it('should invoke predicate once for each checked value', function () {
    var source = hot('-1--2--^-3-4-5-6--7-8--9--|');
    var expected =          '--3---5----7-------|';

    var invoked = 0;
    var predicate = function (x) {
      invoked++;
      return isPrime(x);
    };

    var r = source
      .filter(predicate)
      .do(null, null, function () {
        expect(invoked).toEqual(7);
      });

    expectObservable(r).toBe(expected);
  });
```

Could be a bug in `hot`?

**Comment by kwonoj** [comment]
Seems subscriptionframe `^` is not honored with plain subscription, all elements prior to subscription is pushed.

**Comment by blesh** [comment]
That's because plain subscribe would need to be scheduled at `0` on the test scheduler.

**Comment by kwonoj** [comment]
Yes, assumed so. Think do not need to modify behavior of `testScheduler` or either `hot` for above specific cases, @staltz what do you think?

**Comment by staltz** [comment]
Hmm, yeah, I think using `.do()` to make the assert on `invoked` is fine. And now it's clear that it should have been subscribed on the TestScheduler.

**Comment by blesh** [comment]
Yeah, as a general rule, the only way to know for sure when something will happen is to schedule it on the scheduler.

**Comment by blesh** [comment]
Are these complete now, @staltz? I feel like this got wrapped up after this discussion.

**Comment by staltz** [comment]
@blesh did the missing test get submitted by @luisgabriel ? I can't recall seeing it.

**Comment by luisgabriel** [comment]
@staltz @blesh I'll submit a PR. 

But I think there are some other tests that @staltz pointed out earlier in this thread still missing. Two of them are related to dispose(): 
- https://github.com/Reactive-Extensions/RxJS/blob/master/tests/observable/where.js#L288
- https://github.com/Reactive-Extensions/RxJS/blob/master/tests/observable/where.js#L608

I'm not sure how to replicate these.

**Comment by staltz** [comment]
Yeah, me neither, I tried unsubscribing from within the predicate, but I got a chicken-and-egg subscription problem.

**Comment by kwonoj** [comment]
@luisgabriel @staltz , may I possibly ask if there's remaining task for this issue? Seems most of tests are arrived here.

**Comment by staltz** [comment]
@kwonoj these two tests are not yet in RxJS 5:

- https://github.com/Reactive-Extensions/RxJS/blob/master/tests/observable/where.js#L288
- https://github.com/Reactive-Extensions/RxJS/blob/master/tests/observable/where.js#L608

**Comment by staltz** [comment]
I think these tests will be accomplishable after #716 lands.

**Comment by staltz** [comment]
I sent another PR for more tests on `filter`, but I wasn't able to replicate those two RxJS 4 tests. IMO they are rather irrelevant tests. They aren't filter-specific, they are simply testing whether you can unsubscribe from within a predicate (many other operators have predicates or selectors). It's not worth making filter-specific tests for that. I don't think we will miss anything.

So, from my perspective, this issue can be closed as soon as #755 is merged.

**Comment by luisgabriel** [comment]
I agree with @staltz here.

**Comment by kwonoj** [comment]
Since #755 is landed, I'll close this issue.
